### DISCUSSION TITLE
Node.js reusing one connection & transaction safety

### DISCUSSION THREAD

**Discussion by UROjQ6r80p** [planetscale/discussion#535] [Q&A] [answered]
Is one connection reused across all routes in node.js safe way to use planetscale & transactions?
I mean just basic one node.js server, long running classic express.js application with just ONE planetscale  connection.
Should I use PlanetScale Serverless Driver? Should I create new connection for every transaction?

```
const mysql = require('mysql2/promise');
const connection = await mysql.createConnection({...});

...

// pseudo code
router.get('/first', async () => {
    await connection.query(`START TRANSACTION`);
    await connection.query(...)
    await connection.query(`COMMIT`);
});

router.get('/second', async () => {
    await connection.query(`START TRANSACTION`);
    await connection.query(...)
    await connection.query(`COMMIT`);
});

```

Am I losing something by not creating new connection/using connection pool? Is this going to scale with only one connection like above?

I have seen a comment:

> For example, with PlanetScale/Supabase/Fauna, you can consider them "connectionless" - connecting pooling isn't an issue, as database queries are made over HTTP! They provide a database as an API.

Is that really a case for PlanetScale, so that I can use just one connection and it's "pooled" on the PlanetScale side?

**Comment by jgreet** [comment, marked as answer]
Perhaps you can get away with a single connection if only need to run a single query at a time, but I don't see the benefit in it. If you want to execute multiple queries in parallel, which presumably you will, you probably want to use mysql2's connection pool feature. That way you can execute multiple queries at a time without the overhead of establishing a new connection for every transaction. Picking up an open MySQL connection from the pool will yield a quicker total query time than opening a new connection.

See also https://github.com/sidorares/node-mysql2#using-connection-pools

>> For example, with PlanetScale/Supabase/Fauna, you can consider them "connectionless" - connecting pooling isn't an issue, as database queries are made over HTTP! They provide a database as an API.

> Is that really a case for PlanetScale, so that I can use just one connection and it's "pooled" on the PlanetScale side?

The pooling on our side prevents a database from being overwhelmed by the number of connections which could happen, for example, if you have autoscaling serverless functions that suddenly make hundred or thousands of database client connections at once. It can remove the need for an additional proxy/pool service like proxysql or pgbouncer between the application servers and the database. It won't provide any benefit to an application process using a single MySQL connection.

The main advantage of our [database-js](https://github.com/planetscale/database-js) driver is that it can be used in serverless environments that don't allow traditional database connections. That doesn't seem to be relevant here.

**Reply by UROjQ6r80p** [reply]
So for example production site having 100 concurrent users (Scaler/Scaler Pro plan) I should set connection limit to 100 in the mysql2 driver pool and I'm ready to go?

**Reply by mscoutermarsh** [reply]
Hi @UROjQ6r80p,

For selecting a pool size, I would not recommend basing it on number of concurrent users. Having 100 connections is likely excessive for that number of users. The mysql2 driver _should_ only open connections as needed. But from what I remember, once they are open, they are left open until the process restarts.

Without knowing more about how your application is hosted, it's hard to give exact advice. I'd suggest taking a look at this post by Heroku: https://devcenter.heroku.com/articles/node-concurrency.

Starting with a pool size of 10 is likely plenty per server.

**Reply by UROjQ6r80p** [reply]
Hi @mscoutermarsh 
Thanks a lot, I'll look into that.
It's gonna be cheap VPS with 2gb ram eventually scaled up vertically as needed with just express app, no clusters etc. at the moment.
I will start with pool size of 10 and tune it if needed.
